### ISSUE TITLE
Aborting on shard error when rebooting a node right after schema change

### ISSUE THREAD

**Issue by soyacz** [scylladb/scylladb#12863]
## Issue description

- [x] This issue is a regression.
- [ ] It is unknown if this issue is a regression.

Right after we start c-s stress, VM that runs ScyllaDB is softly rebooted. During ScyllaDB shutdown `aborting on shard` error happens:
```
Feb 14 20:40:29 longevity-5gb-1h-SoftRebootNodeMonk-db-node-249f30ed-1 scylla[5314]:  [shard 0] gossip - My status = NORMAL
Feb 14 20:40:29 longevity-5gb-1h-SoftRebootNodeMonk-db-node-249f30ed-1 scylla[5314]:  [shard 0] gossip - Announcing shutdown
Feb 14 20:40:29 longevity-5gb-1h-SoftRebootNodeMonk-db-node-249f30ed-1 scylla[5314]:  [shard 0] storage_service - Node 10.4.1.211 state jump to normal
Feb 14 20:40:29 longevity-5gb-1h-SoftRebootNodeMonk-db-node-249f30ed-1 scylla[5314]:  [shard 0] gossip - Sending a GossipShutdown to 10.4.3.176 with generation 1676405853
Feb 14 20:40:29 longevity-5gb-1h-SoftRebootNodeMonk-db-node-249f30ed-1 scylla[5314]:  [shard 0] gossip - Sending a GossipShutdown to 10.4.1.56 with generation 1676405853
Feb 14 20:40:30 longevity-5gb-1h-SoftRebootNodeMonk-db-node-249f30ed-1 scylla[5314]:  [shard 0] rpc - client 10.4.1.56:65168 msg_id 2:  exception "gate closed" in no_wait handler ignored
Feb 14 20:40:30 longevity-5gb-1h-SoftRebootNodeMonk-db-node-249f30ed-1 scylla[5314]:  [shard 0] rpc - client 10.4.1.56:65168 msg_id 4:  exception "gate closed" in no_wait handler ignored
Feb 14 20:40:31 longevity-5gb-1h-SoftRebootNodeMonk-db-node-249f30ed-1 scylla[5314]:  [shard 0] rpc - client 10.4.1.56:65168 msg_id 5:  exception "gate closed" in no_wait handler ignored
Feb 14 20:40:31 longevity-5gb-1h-SoftRebootNodeMonk-db-node-249f30ed-1 scylla[5314]:  [shard 0] raft - [5d527a46-c2cb-4642-b3f9-44bd8c81cd0a] applier fiber stopped because state machine was aborted: seastar::abort_requested_exception (abort requested)
Feb 14 20:40:31 longevity-5gb-1h-SoftRebootNodeMonk-db-node-249f30ed-1 scylla[5314]: scylla: raft/server.cc:1133: auto raft::server_impl::applier_fiber()::(anonymous class)::operator()(std::vector<log_entry_ptr> &) const: Assertion `last_idx == _applied_idx + batch.size()' failed.
Feb 14 20:40:31 longevity-5gb-1h-SoftRebootNodeMonk-db-node-249f30ed-1 scylla[5314]: Aborting on shard 0.
Feb 14 20:40:31 longevity-5gb-1h-SoftRebootNodeMonk-db-node-249f30ed-1 scylla[5314]: Backtrace:
Feb 14 20:40:31 longevity-5gb-1h-SoftRebootNodeMonk-db-node-249f30ed-1 scylla[5314]:   0x4f4ae48
Feb 14 20:40:31 longevity-5gb-1h-SoftRebootNodeMonk-db-node-249f30ed-1 scylla[5314]:   0x4f7d282
Feb 14 20:40:31 longevity-5gb-1h-SoftRebootNodeMonk-db-node-249f30ed-1 scylla[5314]:   /opt/scylladb/libreloc/libc.so.6+0x3cb4f
Feb 14 20:40:31 longevity-5gb-1h-SoftRebootNodeMonk-db-node-249f30ed-1 scylla[5314]:   /opt/scylladb/libreloc/libc.so.6+0x8ce7b
Feb 14 20:40:31 longevity-5gb-1h-SoftRebootNodeMonk-db-node-249f30ed-1 scylla[5314]:   /opt/scylladb/libreloc/libc.so.6+0x3caa5
Feb 14 20:40:31 longevity-5gb-1h-SoftRebootNodeMonk-db-node-249f30ed-1 scylla[5314]:   /opt/scylladb/libreloc/libc.so.6+0x267fb
Feb 14 20:40:31 longevity-5gb-1h-SoftRebootNodeMonk-db-node-249f30ed-1 scylla[5314]:   /opt/scylladb/libreloc/libc.so.6+0x2671a
Feb 14 20:40:31 longevity-5gb-1h-SoftRebootNodeMonk-db-node-249f30ed-1 scylla[5314]:   /opt/scylladb/libreloc/libc.so.6+0x35655
Feb 14 20:40:31 longevity-5gb-1h-SoftRebootNodeMonk-db-node-249f30ed-1 scylla[5314]:   0x4042b16
Feb 14 20:40:31 longevity-5gb-1h-SoftRebootNodeMonk-db-node-249f30ed-1 scylla[5314]:   0x406e52f
Feb 14 20:40:31 longevity-5gb-1h-SoftRebootNodeMonk-db-node-249f30ed-1 scylla[5314]:   0x11c4a7a
Feb 14 20:40:31 longevity-5gb-1h-SoftRebootNodeMonk-db-node-249f30ed-1 scylla[5314]:   0x4f5b6c4
Feb 14 20:40:31 longevity-5gb-1h-SoftRebootNodeMonk-db-node-249f30ed-1 scylla[5314]:   0x4f5c947
Feb 14 20:40:31 longevity-5gb-1h-SoftRebootNodeMonk-db-node-249f30ed-1 scylla[5314]:   0x4f5bc89
Feb 14 20:40:31 longevity-5gb-1h-SoftRebootNodeMonk-db-node-249f30ed-1 scylla[5314]:   0x4f026e5
Feb 14 20:40:31 longevity-5gb-1h-SoftRebootNodeMonk-db-node-249f30ed-1 scylla[5314]:   0x4f01858
Feb 14 20:40:31 longevity-5gb-1h-SoftRebootNodeMonk-db-node-249f30ed-1 scylla[5314]:   0x110b7f4
Feb 14 20:40:31 longevity-5gb-1h-SoftRebootNodeMonk-db-node-249f30ed-1 scylla[5314]:   0x110d300
Feb 14 20:40:31 longevity-5gb-1h-SoftRebootNodeMonk-db-node-249f30ed-1 scylla[5314]:   0x1109f2a
Feb 14 20:40:31 longevity-5gb-1h-SoftRebootNodeMonk-db-node-249f30ed-1 scylla[5314]:   /opt/scylladb/libreloc/libc.so.6+0x2750f
Feb 14 20:40:31 longevity-5gb-1h-SoftRebootNodeMonk-db-node-249f30ed-1 scylla[5314]:   /opt/scylladb/libreloc/libc.so.6+0x275c8
Feb 14 20:40:31 longevity-5gb-1h-SoftRebootNodeMonk-db-node-249f30ed-1 scylla[5314]:   0x11080e4
Feb 14 20:41:13 longevity-5gb-1h-SoftRebootNodeMonk-db-node-249f30ed-1 systemd[1]: scylla-server.service: Main process exited, code=dumped, status=6/ABRT
Feb 14 20:41:14 longevity-5gb-1h-SoftRebootNodeMonk-db-node-249f30ed-1 systemd[1]: scylla-server.service: Failed with result 'core-dump'.
Feb 14 20:41:14 longevity-5gb-1h-SoftRebootNodeMonk-db-node-249f30ed-1 systemd[1]: Stopped Scylla Server.
```
decoded:
```
void seastar::backtrace<seastar::backtrace_buffer::append_backtrace()::{lambda(seastar::frame)#1}>(seastar::backtrace_buffer::append_backtrace()::{lambda(seastar::frame)#1}&&) at ./build/release/seastar/./seastar/include/seastar/util/backtrace.hh:59
 (inlined by) seastar::backtrace_buffer::append_backtrace() at ./build/release/seastar/./seastar/src/core/reactor.cc:783
 (inlined by) seastar::print_with_backtrace(seastar::backtrace_buffer&, bool) at ./build/release/seastar/./seastar/src/core/reactor.cc:813
seastar::print_with_backtrace(char const*, bool) at ./build/release/seastar/./seastar/src/core/reactor.cc:825
 (inlined by) seastar::sigabrt_action() at ./build/release/seastar/./seastar/src/core/reactor.cc:3864
 (inlined by) operator() at ./build/release/seastar/./seastar/src/core/reactor.cc:3840
 (inlined by) __invoke at ./build/release/seastar/./seastar/src/core/reactor.cc:3836
?? ??:0
?? ??:0
?? ??:0
?? ??:0
?? ??:0
?? ??:0
operator() at ./raft/server.cc:1133
seastar::future<void> std::__invoke_impl<seastar::future<void>, seastar::internal::variant_visitor<raft::server_impl::applier_fiber()::$_2, raft::server_impl::applier_fiber()::$_3, raft::server_impl::applier_fiber()::$_4>, std::vector<seastar::lw_shared_ptr<raft::log_entry const>, std::allocator<seastar::lw_shared_ptr<raft::log_entry const> > >&>(std::__invoke_other, seastar::internal::variant_visitor<raft::server_impl::applier_fiber()::$_2, raft::server_impl::applier_fiber()::$_3, raft::server_impl::applier_fiber()::$_4>&&, std::vector<seastar::lw_shared_ptr<raft::log_entry const>, std::allocator<seastar::lw_shared_ptr<raft::log_entry const> > >&) at /usr/bin/../lib/gcc/x86_64-redhat-linux/12/../../../../include/c++/12/bits/invoke.h:61
 (inlined by) std::__invoke_result<seastar::internal::variant_visitor<raft::server_impl::applier_fiber()::$_2, raft::server_impl::applier_fiber()::$_3, raft::server_impl::applier_fiber()::$_4>, std::vector<seastar::lw_shared_ptr<raft::log_entry const>, std::allocator<seastar::lw_shared_ptr<raft::log_entry const> > >&>::type std::__invoke<seastar::internal::variant_visitor<raft::server_impl::applier_fiber()::$_2, raft::server_impl::applier_fiber()::$_3, raft::server_impl::applier_fiber()::$_4>, std::vector<seastar::lw_shared_ptr<raft::log_entry const>, std::allocator<seastar::lw_shared_ptr<raft::log_entry const> > >&>(seastar::internal::variant_visitor<raft::server_impl::applier_fiber()::$_2, raft::server_impl::applier_fiber()::$_3, raft::server_impl::applier_fiber()::$_4>&&, std::vector<seastar::lw_shared_ptr<raft::log_entry const>, std::allocator<seastar::lw_shared_ptr<raft::log_entry const> > >&) at /usr/bin/../lib/gcc/x86_64-redhat-linux/12/../../../../include/c++/12/bits/invoke.h:96
 (inlined by) std::__detail::__variant::__gen_vtable_impl<std::__detail::__variant::_Multi_array<std::__detail::__variant::__deduce_visit_result<seastar::future<void> > (*)(seastar::internal::variant_visitor<raft::server_impl::applier_fiber()::$_2, raft::server_impl::applier_fiber()::$_3, raft::server_impl::applier_fiber()::$_4>&&, std::variant<std::vector<seastar::lw_shared_ptr<raft::log_entry const>, std::allocator<seastar::lw_shared_ptr<raft::log_entry const> > >, raft::snapshot_descriptor, raft::server_impl::removed_from_config>&)>, std::integer_sequence<unsigned long, 0ul> >::__visit_invoke(seastar::internal::variant_visitor<raft::server_impl::applier_fiber()::$_2, raft::server_impl::applier_fiber()::$_3, raft::server_impl::applier_fiber()::$_4>&&, std::variant<std::vector<seastar::lw_shared_ptr<raft::log_entry const>, std::allocator<seastar::lw_shared_ptr<raft::log_entry const> > >, raft::snapshot_descriptor, raft::server_impl::removed_from_config>&) at /usr/bin/../lib/gcc/x86_64-redhat-linux/12/../../../../include/c++/12/variant:1031
 (inlined by) decltype(auto) std::__do_visit<std::__detail::__variant::__deduce_visit_result<seastar::future<void> >, seastar::internal::variant_visitor<raft::server_impl::applier_fiber()::$_2, raft::server_impl::applier_fiber()::$_3, raft::server_impl::applier_fiber()::$_4>, std::variant<std::vector<seastar::lw_shared_ptr<raft::log_entry const>, std::allocator<seastar::lw_shared_ptr<raft::log_entry const> > >, raft::snapshot_descriptor, raft::server_impl::removed_from_config>&>(seastar::internal::variant_visitor<raft::server_impl::applier_fiber()::$_2, raft::server_impl::applier_fiber()::$_3, raft::server_impl::applier_fiber()::$_4>&&, std::variant<std::vector<seastar::lw_shared_ptr<raft::log_entry const>, std::allocator<seastar::lw_shared_ptr<raft::log_entry const> > >, raft::snapshot_descriptor, raft::server_impl::removed_from_config>&) at /usr/bin/../lib/gcc/x86_64-redhat-linux/12/../../../../include/c++/12/variant:1788
 (inlined by) std::invoke_result<seastar::internal::variant_visitor<raft::server_impl::applier_fiber()::$_2, raft::server_impl::applier_fiber()::$_3, raft::server_impl::applier_fiber()::$_4>, std::__conditional<is_lvalue_reference_v<std::variant<std::vector<seastar::lw_shared_ptr<raft::log_entry const>, std::allocator<seastar::lw_shared_ptr<raft::log_entry const> > >, raft::snapshot_descriptor, raft::server_impl::removed_from_config>&> >::type<std::variant_alternative<0ul, std::remove_reference<decltype (__variant::__as(std::declval<std::variant<std::vector<seastar::lw_shared_ptr<raft::log_entry const>, std::allocator<seastar::lw_shared_ptr<raft::log_entry const> > >, raft::snapshot_descriptor, raft::server_impl::removed_from_config>&>()))>::type>::type&, std::variant_alternative<0ul, std::remove_reference<decltype (__variant::__as(std::declval<std::variant<std::vector<seastar::lw_shared_ptr<raft::log_entry const>, std::allocator<seastar::lw_shared_ptr<raft::log_entry const> > >, raft::snapshot_descriptor, raft::server_impl::removed_from_config>&>()))>::type>::type&&> >::type std::visit<seastar::internal::variant_visitor<raft::server_impl::applier_fiber()::$_2, raft::server_impl::applier_fiber()::$_3, raft::server_impl::applier_fiber()::$_4>, std::variant<std::vector<seastar::lw_shared_ptr<raft::log_entry const>, std::allocator<seastar::lw_shared_ptr<raft::log_entry const> > >, raft::snapshot_descriptor, raft::server_impl::removed_from_config>&>(seastar::internal::variant_visitor<raft::server_impl::applier_fiber()::$_2, raft::server_impl::applier_fiber()::$_3, raft::server_impl::applier_fiber()::$_4>&&, std::variant<std::vector<seastar::lw_shared_ptr<raft::log_entry const>, std::allocator<seastar::lw_shared_ptr<raft::log_entry const> > >, raft::snapshot_descriptor, raft::server_impl::removed_from_config>&) at /usr/bin/../lib/gcc/x86_64-redhat-linux/12/../../../../include/c++/12/variant:1851
 (inlined by) raft::server_impl::applier_fiber() at ./raft/server.cc:1114
std::__n4861::coroutine_handle<seastar::internal::coroutine_traits_base<void>::promise_type>::resume() const at /usr/bin/../lib/gcc/x86_64-redhat-linux/12/../../../../include/c++/12/coroutine:244
 (inlined by) seastar::internal::coroutine_traits_base<void>::promise_type::run_and_dispose() at ././seastar/include/seastar/core/coroutine.hh:120
seastar::reactor::run_tasks(seastar::reactor::task_queue&) at ./build/release/seastar/./seastar/src/core/reactor.cc:2509
 (inlined by) seastar::reactor::run_some_tasks() at ./build/release/seastar/./seastar/src/core/reactor.cc:2946
seastar::reactor::do_run() at ./build/release/seastar/./seastar/src/core/reactor.cc:3115
seastar::reactor::run() at ./build/release/seastar/./seastar/src/core/reactor.cc:2998
seastar::app_template::run_deprecated(int, char**, std::function<void ()>&&) at ./build/release/seastar/./seastar/src/core/app-template.cc:266
seastar::app_template::run(int, char**, std::function<seastar::future<int> ()>&&) at ./build/release/seastar/./seastar/src/core/app-template.cc:157
scylla_main(int, char**) at ./main.cc:546
std::function<int (int, char**)>::operator()(int, char**) const at /usr/bin/../lib/gcc/x86_64-redhat-linux/12/../../../../include/c++/12/bits/std_function.h:591
main at ./main.cc:1810
?? ??:0
?? ??:0
_start at ??:?
?? ??:0
?? ??:0
?? ??:0
?? ??:0
?? ??:0
?? ??:0
?? ??:0
?? ??:0
?? ??:0
?? ??:0
?? ??:0
?? ??:0
?? ??:0
?? ??:0

```

Unfortunately VM reboot process already killed coredump service and we don't have coredump for it.
```
< t:2023-02-14 20:41:23,979 f:db_log_reader.py l:113  c:sdcm.db_log_reader   p:DEBUG > 2023-02-14T20:41:23+00:00 longevity-5gb-1h-SoftRebootNodeMonk-db-node-249f30ed-1      !ERR | systemd-coredump[8354]: Failed to connect to coredump service: Connection refused
```
from other node logs, it looks c-s schema creation happened when node got shutdown request:
```
Feb 14 20:40:31 longevity-5gb-1h-SoftRebootNodeMonk-db-node-249f30ed-3 scylla[5512]:  [shard 0] schema_tables - Creating keyspace keyspace1
Feb 14 20:40:31 longevity-5gb-1h-SoftRebootNodeMonk-db-node-249f30ed-3 scylla[5512]:  [shard 0] schema_tables - Schema version changed to cd8a75f9-7689-3ee3-8c4f-965954a1319f
```

## Impact

abnormal shutdown, coredump.

## How frequently does it reproduce?

Don't know. I created it as separate issue as I couldn't find any related issue to:
```
scylla: raft/server.cc:1133: auto raft::server_impl::applier_fiber()::(anonymous class)::operator()(std::vector<log_entry_ptr> &) const: Assertion `last_idx == _applied_idx + batch.size()' failed.
```
## Installation details

Kernel Version: 5.15.0-1028-aws
Scylla version (or git commit hash): `5.2.0~rc1-20230207.8ff4717fd010` with build-id `78fbb2c25e9244a62f57988313388a0260084528`

Cluster size: 3 nodes (i4i.large)

Scylla Nodes used in this run:
- longevity-5gb-1h-SoftRebootNodeMonk-db-node-249f30ed-3 (3.252.166.103 | 10.4.3.176) (shards: 2)
- longevity-5gb-1h-SoftRebootNodeMonk-db-node-249f30ed-2 (34.245.61.44 | 10.4.1.56) (shards: 2)
- longevity-5gb-1h-SoftRebootNodeMonk-db-node-249f30ed-1 (34.240.130.155 | 10.4.1.211) (shards: 2)

OS / Image: `ami-05e1d6aa4f71f3f25` (aws: eu-west-1)

Test: `longevity-5gb-1h-SoftRebootNodeMonkey-aws-test`
Test id: `249f30ed-7007-4b8e-a320-1207ebca5e5d`
Test name: `scylla-5.2/nemesis/longevity-5gb-1h-SoftRebootNodeMonkey-aws-test`
Test config file(s):

- [longevity-5gb-1h-SoftRebootNodeMonkey.yaml](https://github.com/scylladb/scylla-cluster-tests/blob/b75ec6c71b99497afbe3bcc89818eadb8d60a9e8/test-cases/nemesis/longevity-5gb-1h-SoftRebootNodeMonkey.yaml)

<details>
<summary>
Logs and commands
</summary>

- Restore Monitor Stack command: `$ hydra investigate show-monitor 249f30ed-7007-4b8e-a320-1207ebca5e5d`
- Restore monitor on AWS instance using [Jenkins job](https://jenkins.scylladb.com/view/QA/job/QA-tools/job/hydra-show-monitor/parambuild/?test_id=249f30ed-7007-4b8e-a320-1207ebca5e5d)
- Show all stored logs command: `$ hydra investigate show-logs 249f30ed-7007-4b8e-a320-1207ebca5e5d`

## Logs:
- **db-cluster-249f30ed.tar.gz** - [https://cloudius-jenkins-test.s3.amazonaws.com/249f30ed-7007-4b8e-a320-1207ebca5e5d/20230214_215652/db-cluster-249f30ed.tar.gz](https://cloudius-jenkins-test.s3.amazonaws.com/249f30ed-7007-4b8e-a320-1207ebca5e5d/20230214_215652/db-cluster-249f30ed.tar.gz)
- **sct-runner-249f30ed.tar.gz** - [https://cloudius-jenkins-test.s3.amazonaws.com/249f30ed-7007-4b8e-a320-1207ebca5e5d/20230214_215652/sct-runner-249f30ed.tar.gz](https://cloudius-jenkins-test.s3.amazonaws.com/249f30ed-7007-4b8e-a320-1207ebca5e5d/20230214_215652/sct-runner-249f30ed.tar.gz)
- **monitor-set-249f30ed.tar.gz** - [https://cloudius-jenkins-test.s3.amazonaws.com/249f30ed-7007-4b8e-a320-1207ebca5e5d/20230214_215652/monitor-set-249f30ed.tar.gz](https://cloudius-jenkins-test.s3.amazonaws.com/249f30ed-7007-4b8e-a320-1207ebca5e5d/20230214_215652/monitor-set-249f30ed.tar.gz)
- **loader-set-249f30ed.tar.gz** - [https://cloudius-jenkins-test.s3.amazonaws.com/249f30ed-7007-4b8e-a320-1207ebca5e5d/20230214_215652/loader-set-249f30ed.tar.gz](https://cloudius-jenkins-test.s3.amazonaws.com/249f30ed-7007-4b8e-a320-1207ebca5e5d/20230214_215652/loader-set-249f30ed.tar.gz)
- **parallel-timelines-report-249f30ed.tar.gz** - [https://cloudius-jenkins-test.s3.amazonaws.com/249f30ed-7007-4b8e-a320-1207ebca5e5d/20230214_215652/parallel-timelines-report-249f30ed.tar.gz](https://cloudius-jenkins-test.s3.amazonaws.com/249f30ed-7007-4b8e-a320-1207ebca5e5d/20230214_215652/parallel-timelines-report-249f30ed.tar.gz)

[Jenkins job URL](https://jenkins.scylladb.com/job/scylla-5.2/job/nemesis/job/longevity-5gb-1h-SoftRebootNodeMonkey-aws-test/4/)
</details>

**Comment by mykaul** [comment]
@kostja - could that be in your team's?

**Comment by kostja** [comment]
it definitely is

**Comment by kostja** [comment]
@gleb-cloudius I checked the code and I don't see how the shutdown event during apply could lead to violation of the invariant above. Perhaps we're concurently applying some outdated snapshot? @kbr-scylla jfyi may be related to https://github.com/scylladb/scylladb/issues/9956

**Comment by gleb-cloudius** [comment]
> @gleb-cloudius I checked the code and I don't see how the shutdown event during apply could lead to violation of the invariant above. Perhaps we're concurently applying some outdated snapshot? @kbr-scylla jfyi may be related to #9956

You are asking yourself a wrong question. The right one is why do we even get to this assert if we cough  `abort_requested_exception` above.

**Comment by kostja** [comment]
During abort, we're waiting for all the futures. Perhaps we should have some abort order.

**Comment by gleb-cloudius** [comment]
> During abort, we're waiting for all the futures concurrently. Perhaps we should have some abort order.

Abort order does not depend on the order you are waiting for futures in and what I pointed above is not related to it anyway. The problem is that apply fiber continues to run after it finds out that the state machine is aborted.

**Comment by gleb-cloudius** [comment]
Something like this should fix it:

```
diff --git a/raft/server.cc b/raft/server.cc
index 4ddd1b54a1..13c53c5971 100644
--- a/raft/server.cc
+++ b/raft/server.cc
@@ -1169,7 +1169,7 @@ future<> server_impl::applier_fiber() {
                         co_await _state_machine->apply(std::move(commands));
                     } catch (abort_requested_exception& e) {
                         logger.info("[{}] applier fiber stopped because state machine was aborted: {}", _id, e);
-                        co_return;
+                        throw stop_apply_fiber{};
                     } catch (...) {
                         std::throw_with_nested(raft::state_machine_error{});
                     }
```

**Comment by kostja** [comment]
Ugh. That's been introduced apparently when we added std::visit. But I still don't get how it can screw the applied_idx.

**Comment by gleb-cloudius** [comment]
> Ugh. That's been introduced apparently when we added std::visit. But I still don't get how it can screw the applied_idx.

By not updating it when abort exception throws and going for the next batch. What is strange here is that the state machine does not throw abort consistently, but OTOH it does not have to.

**Comment by kostja** [comment]
If there is the next batch, the assert happens before the state machine gets a chance to execute.

**Comment by gleb-cloudius** [comment]
> If there is the next batch, the assert happens before the state machine gets a chance to execute.

Ah, right. So nothing strange, just a bug.

**Comment by ShlomiBalalis** [comment]
The issue was reproduced in another 5.2 job:
## Issue description

- [x] This issue is a regression.
- [ ] It is unknown if this issue is a regression.

At `2023-02-14 20:51:04,797` a RestartWithResharding nemesis has started with node 3 as the target node. The nemesis stops scylla, alters the value of `murmur3_partitioner_ignore_msb_bits` in scylla.yaml and starts scylla again, expecting resharding to take place. The node, however, was aborted during the shutdown procedure, after announcing its shutdown:
```
2023-02-14T20:51:06+00:00 longevity-5gb-1h-NodeRestartWithRes-db-node-ffa1bffd-3     !INFO | scylla[5510]:  [shard 0] init - Signal received; shutting down
2023-02-14T20:51:06+00:00 longevity-5gb-1h-NodeRestartWithRes-db-node-ffa1bffd-3     !INFO | scylla[5510]:  [shard 0] init - Shutting down view builder ops
2023-02-14T20:51:06+00:00 longevity-5gb-1h-NodeRestartWithRes-db-node-ffa1bffd-3     !INFO | scylla[5510]:  [shard 0] view - Draining view builder
2023-02-14T20:51:06+00:00 longevity-5gb-1h-NodeRestartWithRes-db-node-ffa1bffd-3     !INFO | scylla[5510]:  [shard 0] compaction_manager - Stopped
2023-02-14T20:51:06+00:00 longevity-5gb-1h-NodeRestartWithRes-db-node-ffa1bffd-3     !INFO | scylla[5510]:  [shard 1] compaction_manager - Asked to stop
2023-02-14T20:51:06+00:00 longevity-5gb-1h-NodeRestartWithRes-db-node-ffa1bffd-3     !INFO | scylla[5510]:  [shard 1] view - Draining view builder
2023-02-14T20:51:06+00:00 longevity-5gb-1h-NodeRestartWithRes-db-node-ffa1bffd-3     !INFO | scylla[5510]:  [shard 1] compaction_manager - Stopped
2023-02-14T20:51:06+00:00 longevity-5gb-1h-NodeRestartWithRes-db-node-ffa1bffd-3     !INFO | scylla[5510]:  [shard 0] init - Shutting down view builder ops was successful
2023-02-14T20:51:06+00:00 longevity-5gb-1h-NodeRestartWithRes-db-node-ffa1bffd-3     !INFO | scylla[5510]:  [shard 0] init - Shutting down local storage
2023-02-14T20:51:06+00:00 longevity-5gb-1h-NodeRestartWithRes-db-node-ffa1bffd-3     !INFO | scylla[5510]:  [shard 0] storage_service - Stop transport: starts
2023-02-14T20:51:06+00:00 longevity-5gb-1h-NodeRestartWithRes-db-node-ffa1bffd-3     !INFO | scylla[5510]:  [shard 0] migration_manager - stopping migration service
2023-02-14T20:51:06+00:00 longevity-5gb-1h-NodeRestartWithRes-db-node-ffa1bffd-3  !WARNING | scylla[5510]:  [shard 0] gossip - failure_detector_loop: Got error in the loop, live_nodes=[10.4.0.193, 10.4.3.34]: seastar::sleep_aborted (Sleep is aborted)
2023-02-14T20:51:06+00:00 longevity-5gb-1h-NodeRestartWithRes-db-node-ffa1bffd-3     !INFO | scylla[5510]:  [shard 0] gossip - failure_detector_loop: Finished main loop
2023-02-14T20:51:06+00:00 longevity-5gb-1h-NodeRestartWithRes-db-node-ffa1bffd-3     !INFO | scylla[5510]:  [shard 1] migration_manager - stopping migration service
2023-02-14T20:51:06+00:00 longevity-5gb-1h-NodeRestartWithRes-db-node-ffa1bffd-3     !INFO | scylla[5510]:  [shard 0] storage_service - Shutting down native transport server
2023-02-14T20:51:06+00:00 longevity-5gb-1h-NodeRestartWithRes-db-node-ffa1bffd-3     !INFO | scylla[5510]:  [shard 0] cql_server_controller - CQL server stopped
2023-02-14T20:51:06+00:00 longevity-5gb-1h-NodeRestartWithRes-db-node-ffa1bffd-3     !INFO | scylla[5510]:  [shard 0] storage_service - Shutting down native transport server was successful
2023-02-14T20:51:06+00:00 longevity-5gb-1h-NodeRestartWithRes-db-node-ffa1bffd-3     !INFO | scylla[5510]:  [shard 0] storage_service - Shutting down rpc server
2023-02-14T20:51:06+00:00 longevity-5gb-1h-NodeRestartWithRes-db-node-ffa1bffd-3     !INFO | scylla[5510]:  [shard 0] storage_service - Shutting down rpc server was successful
2023-02-14T20:51:06+00:00 longevity-5gb-1h-NodeRestartWithRes-db-node-ffa1bffd-3     !INFO | scylla[5510]:  [shard 0] storage_service - Shutting down alternator server
2023-02-14T20:51:06+00:00 longevity-5gb-1h-NodeRestartWithRes-db-node-ffa1bffd-3     !INFO | scylla[5510]:  [shard 0] storage_service - Shutting down alternator server was successful
2023-02-14T20:51:06+00:00 longevity-5gb-1h-NodeRestartWithRes-db-node-ffa1bffd-3     !INFO | scylla[5510]:  [shard 0] storage_service - Shutting down redis server
2023-02-14T20:51:06+00:00 longevity-5gb-1h-NodeRestartWithRes-db-node-ffa1bffd-3     !INFO | scylla[5510]:  [shard 0] storage_service - Shutting down redis server was successful
2023-02-14T20:51:06+00:00 longevity-5gb-1h-NodeRestartWithRes-db-node-ffa1bffd-3     !INFO | scylla[5510]:  [shard 0] storage_service - Stop transport: shutdown rpc and cql server done
2023-02-14T20:51:06+00:00 longevity-5gb-1h-NodeRestartWithRes-db-node-ffa1bffd-3     !INFO | scylla[5510]:  [shard 0] gossip - My status = NORMAL
2023-02-14T20:51:06+00:00 longevity-5gb-1h-NodeRestartWithRes-db-node-ffa1bffd-3     !INFO | scylla[5510]:  [shard 0] gossip - Announcing shutdown
2023-02-14T20:51:06+00:00 longevity-5gb-1h-NodeRestartWithRes-db-node-ffa1bffd-3     !INFO | scylla[5510]:  [shard 0] storage_service - Node 10.4.0.12 state jump to normal
2023-02-14T20:51:06+00:00 longevity-5gb-1h-NodeRestartWithRes-db-node-ffa1bffd-3     !INFO | scylla[5510]:  [shard 0] gossip - Sending a GossipShutdown to 10.4.0.193 with generation 1676406940
2023-02-14T20:51:06+00:00 longevity-5gb-1h-NodeRestartWithRes-db-node-ffa1bffd-3     !INFO | scylla[5510]:  [shard 0] gossip - Sending a GossipShutdown to 10.4.3.34 with generation 1676406940
2023-02-14T20:51:07+00:00 longevity-5gb-1h-NodeRestartWithRes-db-node-ffa1bffd-3     !INFO | scylla[5510]:  [shard 0] rpc - client 10.4.3.34:52498 msg_id 2:  exception "gate closed" in no_wait handler ignored
2023-02-14T20:51:07+00:00 longevity-5gb-1h-NodeRestartWithRes-db-node-ffa1bffd-3     !INFO | scylla[5510]:  [shard 0] rpc - client 10.4.3.34:52498 msg_id 3:  exception "gate closed" in no_wait handler ignored
2023-02-14T20:51:07+00:00 longevity-5gb-1h-NodeRestartWithRes-db-node-ffa1bffd-3     !INFO | scylla[5510]:  [shard 0] rpc - client 10.4.0.193:55092 msg_id 2:  exception "gate closed" in no_wait handler ignored
2023-02-14T20:51:07+00:00 longevity-5gb-1h-NodeRestartWithRes-db-node-ffa1bffd-3     !INFO | scylla[5510]:  [shard 0] rpc - client 10.4.0.193:55092 msg_id 3:  exception "gate closed" in no_wait handler ignored
2023-02-14T20:51:08+00:00 longevity-5gb-1h-NodeRestartWithRes-db-node-ffa1bffd-3     !INFO | scylla[5510]:  [shard 0] rpc - client 10.4.3.34:52498 msg_id 4:  exception "gate closed" in no_wait handler ignored
2023-02-14T20:51:08+00:00 longevity-5gb-1h-NodeRestartWithRes-db-node-ffa1bffd-3     !INFO | scylla[5510]:  [shard 0] raft - [0f68d2d5-40ce-47a6-b9ae-dc35326f0ed3] applier fiber stopped because state machine was aborted: seastar::abort_requested_exception (abort requested)
2023-02-14T20:51:08+00:00 longevity-5gb-1h-NodeRestartWithRes-db-node-ffa1bffd-3     !INFO | scylla[5510]:  [shard 0] rpc - client 10.4.3.34:52498 msg_id 5:  exception "gate closed" in no_wait handler ignored
2023-02-14T20:51:08+00:00 longevity-5gb-1h-NodeRestartWithRes-db-node-ffa1bffd-3     !INFO | scylla[5510]:  [shard 0] rpc - client 10.4.0.193:55092 msg_id 4:  exception "gate closed" in no_wait handler ignored
2023-02-14T20:51:08+00:00 longevity-5gb-1h-NodeRestartWithRes-db-node-ffa1bffd-3     !INFO | scylla[5510]: scylla: raft/server.cc:1133: auto raft::server_impl::applier_fiber()::(anonymous class)::operator()(std::vector<log_entry_ptr> &) const: Assertion `last_idx == _applied_idx + batch.size()' failed.
2023-02-14T20:51:08+00:00 longevity-5gb-1h-NodeRestartWithRes-db-node-ffa1bffd-3     !INFO | scylla[5510]: Aborting on shard 0.
2023-02-14T20:51:08+00:00 longevity-5gb-1h-NodeRestartWithRes-db-node-ffa1bffd-3     !INFO | scylla[5510]: Backtrace:
2023-02-14T20:51:08+00:00 longevity-5gb-1h-NodeRestartWithRes-db-node-ffa1bffd-3     !INFO | scylla[5510]:   0x4f4ae48
2023-02-14T20:51:08+00:00 longevity-5gb-1h-NodeRestartWithRes-db-node-ffa1bffd-3     !INFO | scylla[5510]:   0x4f7d282
2023-02-14T20:51:08+00:00 longevity-5gb-1h-NodeRestartWithRes-db-node-ffa1bffd-3     !INFO | scylla[5510]:   /opt/scylladb/libreloc/libc.so.6+0x3cb4f
2023-02-14T20:51:08+00:00 longevity-5gb-1h-NodeRestartWithRes-db-node-ffa1bffd-3     !INFO | scylla[5510]:   /opt/scylladb/libreloc/libc.so.6+0x8ce7b
2023-02-14T20:51:08+00:00 longevity-5gb-1h-NodeRestartWithRes-db-node-ffa1bffd-3     !INFO | scylla[5510]:   /opt/scylladb/libreloc/libc.so.6+0x3caa5
2023-02-14T20:51:08+00:00 longevity-5gb-1h-NodeRestartWithRes-db-node-ffa1bffd-3     !INFO | scylla[5510]:   /opt/scylladb/libreloc/libc.so.6+0x267fb
2023-02-14T20:51:08+00:00 longevity-5gb-1h-NodeRestartWithRes-db-node-ffa1bffd-3     !INFO | scylla[5510]:   /opt/scylladb/libreloc/libc.so.6+0x2671a
2023-02-14T20:51:08+00:00 longevity-5gb-1h-NodeRestartWithRes-db-node-ffa1bffd-3     !INFO | scylla[5510]:   /opt/scylladb/libreloc/libc.so.6+0x35655
2023-02-14T20:51:08+00:00 longevity-5gb-1h-NodeRestartWithRes-db-node-ffa1bffd-3     !INFO | scylla[5510]:   0x4042b16
2023-02-14T20:51:08+00:00 longevity-5gb-1h-NodeRestartWithRes-db-node-ffa1bffd-3     !INFO | scylla[5510]:   0x406e52f
2023-02-14T20:51:08+00:00 longevity-5gb-1h-NodeRestartWithRes-db-node-ffa1bffd-3     !INFO | scylla[5510]:   0x11c4a7a
2023-02-14T20:51:08+00:00 longevity-5gb-1h-NodeRestartWithRes-db-node-ffa1bffd-3     !INFO | scylla[5510]:   0x4f5b6c4
2023-02-14T20:51:08+00:00 longevity-5gb-1h-NodeRestartWithRes-db-node-ffa1bffd-3     !INFO | scylla[5510]:   0x4f5c947
2023-02-14T20:51:08+00:00 longevity-5gb-1h-NodeRestartWithRes-db-node-ffa1bffd-3     !INFO | scylla[5510]:   0x4f5bc89
2023-02-14T20:51:08+00:00 longevity-5gb-1h-NodeRestartWithRes-db-node-ffa1bffd-3     !INFO | scylla[5510]:   0x4f026e5
2023-02-14T20:51:08+00:00 longevity-5gb-1h-NodeRestartWithRes-db-node-ffa1bffd-3     !INFO | scylla[5510]:   0x4f01858
2023-02-14T20:51:08+00:00 longevity-5gb-1h-NodeRestartWithRes-db-node-ffa1bffd-3     !INFO | scylla[5510]:   0x110b7f4
2023-02-14T20:51:08+00:00 longevity-5gb-1h-NodeRestartWithRes-db-node-ffa1bffd-3     !INFO | scylla[5510]:   0x110d300
2023-02-14T20:51:08+00:00 longevity-5gb-1h-NodeRestartWithRes-db-node-ffa1bffd-3     !INFO | scylla[5510]:   0x1109f2a
2023-02-14T20:51:08+00:00 longevity-5gb-1h-NodeRestartWithRes-db-node-ffa1bffd-3     !INFO | scylla[5510]:   /opt/scylladb/libreloc/libc.so.6+0x2750f
2023-02-14T20:51:08+00:00 longevity-5gb-1h-NodeRestartWithRes-db-node-ffa1bffd-3     !INFO | scylla[5510]:   /opt/scylladb/libreloc/libc.so.6+0x275c8
2023-02-14T20:51:08+00:00 longevity-5gb-1h-NodeRestartWithRes-db-node-ffa1bffd-3     !INFO | scylla[5510]:   0x11080e4
```
Translated backtrace:
```
2023-02-14T20:51:08+00:00 longevity-5gb-1h-NodeRestartWithRes-db-node-ffa1bffd-3     !INFO | scylla[5510]: Aborting on shard 0.
void seastar::backtrace<seastar::backtrace_buffer::append_backtrace()::{lambda(seastar::frame)#1}>(seastar::backtrace_buffer::append_backtrace()::{lambda(seastar::frame)#1}&&) at ./build/release/seastar/./seastar/include/seastar/util/backtrace.hh:59
 (inlined by) seastar::backtrace_buffer::append_backtrace() at ./build/release/seastar/./seastar/src/core/reactor.cc:783
 (inlined by) seastar::print_with_backtrace(seastar::backtrace_buffer&, bool) at ./build/release/seastar/./seastar/src/core/reactor.cc:813
seastar::print_with_backtrace(char const*, bool) at ./build/release/seastar/./seastar/src/core/reactor.cc:825
 (inlined by) seastar::sigabrt_action() at ./build/release/seastar/./seastar/src/core/reactor.cc:3864
 (inlined by) operator() at ./build/release/seastar/./seastar/src/core/reactor.cc:3840
 (inlined by) __invoke at ./build/release/seastar/./seastar/src/core/reactor.cc:3836
?? ??:0
?? ??:0
?? ??:0
?? ??:0
?? ??:0
?? ??:0
operator() at ./raft/server.cc:1133
seastar::future<void> std::__invoke_impl<seastar::future<void>, seastar::internal::variant_visitor<raft::server_impl::applier_fiber()::$_2, raft::server_impl::applier_fiber()::$_3, raft::server_impl::applier_fiber()::$_4>, std::vector<seastar::lw_shared_ptr<raft::log_entry const>, std::allocator<seastar::lw_shared_ptr<raft::log_entry const> > >&>(std::__invoke_other, seastar::internal::variant_visitor<raft::server_impl::applier_fiber()::$_2, raft::server_impl::applier_fiber()::$_3, raft::server_impl::applier_fiber()::$_4>&&, std::vector<seastar::lw_shared_ptr<raft::log_entry const>, std::allocator<seastar::lw_shared_ptr<raft::log_entry const> > >&) at /usr/bin/../lib/gcc/x86_64-redhat-linux/12/../../../../include/c++/12/bits/invoke.h:61
 (inlined by) std::__invoke_result<seastar::internal::variant_visitor<raft::server_impl::applier_fiber()::$_2, raft::server_impl::applier_fiber()::$_3, raft::server_impl::applier_fiber()::$_4>, std::vector<seastar::lw_shared_ptr<raft::log_entry const>, std::allocator<seastar::lw_shared_ptr<raft::log_entry const> > >&>::type std::__invoke<seastar::internal::variant_visitor<raft::server_impl::applier_fiber()::$_2, raft::server_impl::applier_fiber()::$_3, raft::server_impl::applier_fiber()::$_4>, std::vector<seastar::lw_shared_ptr<raft::log_entry const>, std::allocator<seastar::lw_shared_ptr<raft::log_entry const> > >&>(seastar::internal::variant_visitor<raft::server_impl::applier_fiber()::$_2, raft::server_impl::applier_fiber()::$_3, raft::server_impl::applier_fiber()::$_4>&&, std::vector<seastar::lw_shared_ptr<raft::log_entry const>, std::allocator<seastar::lw_shared_ptr<raft::log_entry const> > >&) at /usr/bin/../lib/gcc/x86_64-redhat-linux/12/../../../../include/c++/12/bits/invoke.h:96
 (inlined by) std::__detail::__variant::__gen_vtable_impl<std::__detail::__variant::_Multi_array<std::__detail::__variant::__deduce_visit_result<seastar::future<void> > (*)(seastar::internal::variant_visitor<raft::server_impl::applier_fiber()::$_2, raft::server_impl::applier_fiber()::$_3, raft::server_impl::applier_fiber()::$_4>&&, std::variant<std::vector<seastar::lw_shared_ptr<raft::log_entry const>, std::allocator<seastar::lw_shared_ptr<raft::log_entry const> > >, raft::snapshot_descriptor, raft::server_impl::removed_from_config>&)>, std::integer_sequence<unsigned long, 0ul> >::__visit_invoke(seastar::internal::variant_visitor<raft::server_impl::applier_fiber()::$_2, raft::server_impl::applier_fiber()::$_3, raft::server_impl::applier_fiber()::$_4>&&, std::variant<std::vector<seastar::lw_shared_ptr<raft::log_entry const>, std::allocator<seastar::lw_shared_ptr<raft::log_entry const> > >, raft::snapshot_descriptor, raft::server_impl::removed_from_config>&) at /usr/bin/../lib/gcc/x86_64-redhat-linux/12/../../../../include/c++/12/variant:1031
 (inlined by) decltype(auto) std::__do_visit<std::__detail::__variant::__deduce_visit_result<seastar::future<void> >, seastar::internal::variant_visitor<raft::server_impl::applier_fiber()::$_2, raft::server_impl::applier_fiber()::$_3, raft::server_impl::applier_fiber()::$_4>, std::variant<std::vector<seastar::lw_shared_ptr<raft::log_entry const>, std::allocator<seastar::lw_shared_ptr<raft::log_entry const> > >, raft::snapshot_descriptor, raft::server_impl::removed_from_config>&>(seastar::internal::variant_visitor<raft::server_impl::applier_fiber()::$_2, raft::server_impl::applier_fiber()::$_3, raft::server_impl::applier_fiber()::$_4>&&, std::variant<std::vector<seastar::lw_shared_ptr<raft::log_entry const>, std::allocator<seastar::lw_shared_ptr<raft::log_entry const> > >, raft::snapshot_descriptor, raft::server_impl::removed_from_config>&) at /usr/bin/../lib/gcc/x86_64-redhat-linux/12/../../../../include/c++/12/variant:1788
 (inlined by) std::invoke_result<seastar::internal::variant_visitor<raft::server_impl::applier_fiber()::$_2, raft::server_impl::applier_fiber()::$_3, raft::server_impl::applier_fiber()::$_4>, std::__conditional<is_lvalue_reference_v<std::variant<std::vector<seastar::lw_shared_ptr<raft::log_entry const>, std::allocator<seastar::lw_shared_ptr<raft::log_entry const> > >, raft::snapshot_descriptor, raft::server_impl::removed_from_config>&> >::type<std::variant_alternative<0ul, std::remove_reference<decltype (__variant::__as(std::declval<std::variant<std::vector<seastar::lw_shared_ptr<raft::log_entry const>, std::allocator<seastar::lw_shared_ptr<raft::log_entry const> > >, raft::snapshot_descriptor, raft::server_impl::removed_from_config>&>()))>::type>::type&, std::variant_alternative<0ul, std::remove_reference<decltype (__variant::__as(std::declval<std::variant<std::vector<seastar::lw_shared_ptr<raft::log_entry const>, std::allocator<seastar::lw_shared_ptr<raft::log_entry const> > >, raft::snapshot_descriptor, raft::server_impl::removed_from_config>&>()))>::type>::type&&> >::type std::visit<seastar::internal::variant_visitor<raft::server_impl::applier_fiber()::$_2, raft::server_impl::applier_fiber()::$_3, raft::server_impl::applier_fiber()::$_4>, std::variant<std::vector<seastar::lw_shared_ptr<raft::log_entry const>, std::allocator<seastar::lw_shared_ptr<raft::log_entry const> > >, raft::snapshot_descriptor, raft::server_impl::removed_from_config>&>(seastar::internal::variant_visitor<raft::server_impl::applier_fiber()::$_2, raft::server_impl::applier_fiber()::$_3, raft::server_impl::applier_fiber()::$_4>&&, std::variant<std::vector<seastar::lw_shared_ptr<raft::log_entry const>, std::allocator<seastar::lw_shared_ptr<raft::log_entry const> > >, raft::snapshot_descriptor, raft::server_impl::removed_from_config>&) at /usr/bin/../lib/gcc/x86_64-redhat-linux/12/../../../../include/c++/12/variant:1851
 (inlined by) raft::server_impl::applier_fiber() at ./raft/server.cc:1114
std::__n4861::coroutine_handle<seastar::internal::coroutine_traits_base<void>::promise_type>::resume() const at /usr/bin/../lib/gcc/x86_64-redhat-linux/12/../../../../include/c++/12/coroutine:244
 (inlined by) seastar::internal::coroutine_traits_base<void>::promise_type::run_and_dispose() at ././seastar/include/seastar/core/coroutine.hh:120
seastar::reactor::run_tasks(seastar::reactor::task_queue&) at ./build/release/seastar/./seastar/src/core/reactor.cc:2509
 (inlined by) seastar::reactor::run_some_tasks() at ./build/release/seastar/./seastar/src/core/reactor.cc:2946
seastar::reactor::do_run() at ./build/release/seastar/./seastar/src/core/reactor.cc:3115
seastar::reactor::run() at ./build/release/seastar/./seastar/src/core/reactor.cc:2998
seastar::app_template::run_deprecated(int, char**, std::function<void ()>&&) at ./build/release/seastar/./seastar/src/core/app-template.cc:266
seastar::app_template::run(int, char**, std::function<seastar::future<int> ()>&&) at ./build/release/seastar/./seastar/src/core/app-template.cc:157
scylla_main(int, char**) at ./main.cc:546
std::function<int (int, char**)>::operator()(int, char**) const at /usr/bin/../lib/gcc/x86_64-redhat-linux/12/../../../../include/c++/12/bits/std_function.h:591
main at ./main.cc:1810
?? ??:0
?? ??:0
_start at ??:?
```
It seems that the cause for the abortion is Raft, according to:
`2023-02-14T20:51:08+00:00 longevity-5gb-1h-NodeRestartWithRes-db-node-ffa1bffd-3     !INFO | scylla[5510]:  [shard 0] raft - [0f68d2d5-40ce-47a6-b9ae-dc35326f0ed3] applier fiber stopped because state machine was aborted: seastar::abort_requested_exception (abort requested)`

A coredump was also created:
```
backtrace=           PID: 5510 (scylla)
           UID: 112 (scylla)
           GID: 118 (scylla)
        Signal: 6 (ABRT)
     Timestamp: Tue 2023-02-14 20:51:08 UTC (1min 24s ago)
  Command Line: /usr/bin/scylla --blocked-reactor-notify-ms 25 --abort-on-lsa-bad-alloc 1 --abort-on-seastar-bad-alloc --abort-on-internal-error 1 --abort-on-ebadf 1 --enable-sstable-key-validation 1 --log-to-syslog 1 --log-to-stdout 0 --default-log-level info --network-stack posix --io-properties-file=/etc/scylla.d/io_properties.yaml --cpuset 0-1 --lock-memory=1
    Executable: /opt/scylladb/libexec/scylla
 Control Group: /scylla.slice/scylla-server.slice/scylla-server.service
          Unit: scylla-server.service
         Slice: scylla-server.slice
       Boot ID: c5a0c1cbc1104c23ba1802bb59cfe68f
    Machine ID: 42060207b1cd49fe8c8643abb7349d74
      Hostname: longevity-5gb-1h-NodeRestartWithRes-db-node-ffa1bffd-3
       Storage: /var/lib/systemd/coredump/core.scylla.112.c5a0c1cbc1104c23ba1802bb59cfe68f.5510.1676407868000000 (present)
     Disk Size: 13.8G
       Message: Process 5510 (scylla) of user 112 dumped core.
                
                Found module linux-vdso.so.1 with build-id: 7b2a39d3eed27929943a3232d1fde693d4f59680
                Found module libcrypto.so.3 with build-id: 7470c53208efe3360d43315b439f03d64569b15b
                Metadata for module libcrypto.so.3 owned by FDO found: {
                        "type" : "rpm",
                        "name" : "openssl",
                        "version" : "3.0.5-3.fc37",
                        "architecture" : "x86_64",
                        "osCpe" : "cpe:/o:fedoraproject:fedora:37"
                }             
...
                Stack trace of thread 5511:
                #0  0x00007f795ad2a76d syscall (libc.so.6 + 0x10476d)
                #1  0x0000000004fa5121 _ZN7seastar8internal13io_pgeteventsEmllPNS0_9linux_abi8io_eventEPK8timespecPK10__sigset_tb (scylla + 0x4da5121)
                #2  0x0000000004fa0d55 _ZN7seastar19reactor_backend_aio12await_eventsEiPK10__sigset_t (scylla + 0x4da0d55)
                #3  0x0000000004fa1470 _ZN7seastar19reactor_backend_aio23wait_and_process_eventsEPK10__sigset_t (scylla + 0x4da1470)
                #4  0x0000000004f5cc4d _ZN7seastar7reactor6do_runEv (scylla + 0x4d5cc4d)
                #5  0x0000000004f7d732 _ZNSt17_Function_handlerIFvvEZN7seastar3smp9configureERKNS1_11smp_optionsERKNS1_15reactor_optionsEE4$_94E9_M_invokeERKSt9_Any_data (scylla + 0x4d7d732)
                #6  0x0000000004f2da2b _ZN7seastar12posix_thread13start_routineEPv (scylla + 0x4d2da2b)
                #7  0x00007f795acb114d start_thread (libc.so.6 + 0x8b14d)
                #8  0x00007f795ad32a00 __clone3 (libc.so.6 + 0x10ca00)

download_instructions=gsutil cp gs://upload.scylladb.com/core.scylla.112.c5a0c1cbc1104c23ba1802bb59cfe68f.5510.1676407868000000/core.scylla.112.c5a0c1cbc1104c23ba1802bb59cfe68f.5510.1676407868000000.gz .
gunzip /var/lib/systemd/coredump/core.scylla.112.c5a0c1cbc1104c23ba1802bb59cfe68f.5510.1676407868000000.gz
```

After the the node was aborted, we started the scylla service again and it started with no issue (resharding has even finished successfully)

## Impact

While the shutdown did cause the node to abort, the node started up again without issue. So I'd say that the impact is somewhat severe.

## How frequently does it reproduce?

In this run, the nemesis that triggered this issue ran three times, and the issue occurred one out of those three runs. However, node 3 was only the target node only once. In the other two nemesis runs, node 2 was restarted and it did not experience any issue.

## Installation details

Kernel Version: 5.15.0-1028-aws
Scylla version (or git commit hash): `5.2.0~rc1-20230207.8ff4717fd010` with build-id `78fbb2c25e9244a62f57988313388a0260084528`




Cluster size: 3 nodes (i4i.large)

Scylla Nodes used in this run:
- longevity-5gb-1h-NodeRestartWithRes-db-node-ffa1bffd-3 (34.249.9.154 | 10.4.0.12) (shards: 2)
- longevity-5gb-1h-NodeRestartWithRes-db-node-ffa1bffd-2 (54.154.29.252 | 10.4.0.193) (shards: 2)
- longevity-5gb-1h-NodeRestartWithRes-db-node-ffa1bffd-1 (34.250.20.71 | 10.4.3.34) (shards: 2)


OS / Image: `ami-05e1d6aa4f71f3f25` (aws: eu-west-1)

Test: `longevity-5gb-1h-NodeRestartWithResharding-aws-test`
Test id: `ffa1bffd-9302-4ab7-a0ed-3328bcdf2a60`
Test name: `scylla-5.2/nemesis/longevity-5gb-1h-NodeRestartWithResharding-aws-test`
Test config file(s):

- [longevity-5gb-1h-NodeRestartWithResharding.yaml](https://github.com/scylladb/scylla-cluster-tests/blob/b75ec6c71b99497afbe3bcc89818eadb8d60a9e8/test-cases/nemesis/longevity-5gb-1h-NodeRestartWithResharding.yaml)


<details>
<summary>
Logs and commands
</summary>


- Restore Monitor Stack command: `$ hydra investigate show-monitor ffa1bffd-9302-4ab7-a0ed-3328bcdf2a60`
- Restore monitor on AWS instance using [Jenkins job](https://jenkins.scylladb.com/view/QA/job/QA-tools/job/hydra-show-monitor/parambuild/?test_id=ffa1bffd-9302-4ab7-a0ed-3328bcdf2a60)
- Show all stored logs command: `$ hydra investigate show-logs ffa1bffd-9302-4ab7-a0ed-3328bcdf2a60`


## Logs:
- **db-cluster-ffa1bffd.tar.gz** - [https://cloudius-jenkins-test.s3.amazonaws.com/ffa1bffd-9302-4ab7-a0ed-3328bcdf2a60/20230214_220647/db-cluster-ffa1bffd.tar.gz](https://cloudius-jenkins-test.s3.amazonaws.com/ffa1bffd-9302-4ab7-a0ed-3328bcdf2a60/20230214_220647/db-cluster-ffa1bffd.tar.gz)
- **sct-runner-ffa1bffd.tar.gz** - [https://cloudius-jenkins-test.s3.amazonaws.com/ffa1bffd-9302-4ab7-a0ed-3328bcdf2a60/20230214_220647/sct-runner-ffa1bffd.tar.gz](https://cloudius-jenkins-test.s3.amazonaws.com/ffa1bffd-9302-4ab7-a0ed-3328bcdf2a60/20230214_220647/sct-runner-ffa1bffd.tar.gz)
- **monitor-set-ffa1bffd.tar.gz** - [https://cloudius-jenkins-test.s3.amazonaws.com/ffa1bffd-9302-4ab7-a0ed-3328bcdf2a60/20230214_220647/monitor-set-ffa1bffd.tar.gz](https://cloudius-jenkins-test.s3.amazonaws.com/ffa1bffd-9302-4ab7-a0ed-3328bcdf2a60/20230214_220647/monitor-set-ffa1bffd.tar.gz)
- **loader-set-ffa1bffd.tar.gz** - [https://cloudius-jenkins-test.s3.amazonaws.com/ffa1bffd-9302-4ab7-a0ed-3328bcdf2a60/20230214_220647/loader-set-ffa1bffd.tar.gz](https://cloudius-jenkins-test.s3.amazonaws.com/ffa1bffd-9302-4ab7-a0ed-3328bcdf2a60/20230214_220647/loader-set-ffa1bffd.tar.gz)
- **parallel-timelines-report-ffa1bffd.tar.gz** - [https://cloudius-jenkins-test.s3.amazonaws.com/ffa1bffd-9302-4ab7-a0ed-3328bcdf2a60/20230214_220647/parallel-timelines-report-ffa1bffd.tar.gz](https://cloudius-jenkins-test.s3.amazonaws.com/ffa1bffd-9302-4ab7-a0ed-3328bcdf2a60/20230214_220647/parallel-timelines-report-ffa1bffd.tar.gz)


[Jenkins job URL](https://jenkins.scylladb.com/job/scylla-5.2/job/nemesis/job/longevity-5gb-1h-NodeRestartWithResharding-aws-test/4/)
</details>

**Comment by kostja** [comment]
@alecco apart from the fix itself, which gleb sent above, this needs a new nemesis in the randomized_nemesis_test - shutdown nemesis. It can be similar to the crash nemesis, but should inject abort exception into the impure state machine. Could you take this as a high priority item, sending gleb's fix first and then following up with a test? apparently it has a high impact on QA efforts and will need to be backported.
@kbr-scylla please assist with the code review of the nemesis test fixes.

**Comment by kbr-scylla** [comment]
> @alecco apart from the fix itself, which gleb sent above, this needs a new nemesis in the randomized_nemesis_test - shutdown nemesis. It can be similar to the crash nemesis, but should inject abort exception into the impure state machine.

The `crash_stop` nemesis is already simulating both crashes and normal aborts.

Why would we inject anything to abort? We should do the same thing that production code does. Production code does not inject an exception, it calls abort() on the services.

**Comment by kbr-scylla** [comment]
This is the implementation of `group0_state_machine::abort()`:
```cpp
future<> group0_state_machine::abort() {
    return make_ready_future<>();
}
```

and the implementation of `impure_state_machine::abort()` in nemesis tests:
```cpp
    future<> abort() override {
        return _gate.close();
    }
```

Hmm, but perhaps `group0_state_machine::apply()` throws an exception from some underlying service it calls during shutdown?

Even if we wanted to simulate that, we don't need any additional nemeses. It should be enough to modify `raft_server::abort()` function (`raft_server` is a template in randomized_nemesis_test.cc) so it injects the exception into `impure_state_machine`. The `abort()` function is being called by the existing `stop_crash` nemesis.

**Comment by kostja** [comment]
The flow of events in this issue is that state_machine::apply() throws once the machine was aborted. we don't have it in the nemesis test.

What I suggest is to add a shutdown nemesis which will set impure_state_machine::shutdown flag to true.
if this flag is set, the state machien will begin throwing abort_requested exception, just like the current scylla code.
after a while, this state machine will be crashed/destroyed/recreated just like the crash nemesis does.

**Comment by kostja** [comment]
> This is the implementation of `group0_state_machine::abort()`:
> 
> ```c++
> future<> group0_state_machine::abort() {
>     return make_ready_future<>();
> }
> ```
> 
> and the implementation of `impure_state_machine::abort()` in nemesis tests:
> 
> ```c++
>     future<> abort() override {
>         return _gate.close();
>     }
> ```
> 
> Hmm, but perhaps `group0_state_machine::apply()` throws an exception from some underlying service it calls during shutdown?
> 
> Even if we wanted to simulate that, we don't need any additional nemeses. It should be enough to modify `raft_server::abort()` function (`raft_server` is a template in randomized_nemesis_test.cc) so it injects the exception into `impure_state_machine`. The `abort()` function is being called by the existing `stop_crash` nemesis.

Agree to that, but it should not only throw, it should sleep after throwing a bit, or keep the machine running, to give the bug a chance to reproduce.

**Comment by avikivity** [comment]
Backported to 5.2. 5.1 is different here, and raft is experimental there anyway.
